### PR TITLE
Get mock instance

### PR DESCRIPTION
**Pull request by jubianchi**

### CODE REVIEW

File: classes/test.php [hunk 1]
```diff
@@ -411,6 +411,27 @@ public function setAssertionManager(test\assertion\manager $assertionManager = n
 			->setHandler('mockGenerator', function() use ($test) { return $test->getMockGenerator(); })
 			->setHandler('mockClass', function($class, $mockNamespace = null, $mockClass = null) use ($test) { $test->getMockGenerator()->generate($class, $mockNamespace, $mockClass); return $test; })
 			->setHandler('mockTestedClass', function($mockNamespace = null, $mockClass = null) use ($test) { $test->getMockGenerator()->generate($test->getTestedClassName(), $mockNamespace, $mockClass); return $test; })
+			->setHandler('newMockInstance', function($class, $mockNamespace = null, $mockClass = null, array $constructorArguments = null) use ($test)
```
Comment (jubianchi) [inline classes/test.php]: @Grummfy I applied some refactoring here...

File: classes/test.php [hunk 2]
```diff
@@ -411,6 +411,27 @@ public function setAssertionManager(test\assertion\manager $assertionManager = n
...
+				{
+					$mockNamespace = trim($mockNamespace ?: $test->getMockGenerator()->getDefaultNamespace(), '\\');
+					$mockClass = trim($mockClass ?: $class, '\\');
+					$className = $mockNamespace . '\\' . $mockClass;
+
+					if (class_exists($className) === false)
```
Comment (jubianchi) [inline classes/test.php]: we have to check if the class exists before generating it as the mock generator does not check that for us

File: classes/test.php [hunk 3]
```diff
@@ -411,6 +411,27 @@ public function setAssertionManager(test\assertion\manager $assertionManager = n
...
+					{
+						$test->mockClass($class, $mockNamespace, $mockClass);
```
Comment (jubianchi) [inline classes/test.php]: here we use a previously declared handler to generate the mock class

File: classes/test.php [hunk 4]
```diff
@@ -411,6 +411,27 @@ public function setAssertionManager(test\assertion\manager $assertionManager = n
...
+					}
+
+					if ($constructorArguments !== null)
```
Comment (jubianchi) [inline classes/test.php]: we only rely on reflection if needed

File: classes/test.php [hunk 5]
```diff
@@ -411,6 +411,27 @@ public function setAssertionManager(test\assertion\manager $assertionManager = n
...
+					{
+						$reflectionClass = new \reflectionClass($className);
+
+						return $reflectionClass->newInstanceArgs($constructorArguments);
+					}
+
+					return new $className();
+				}
```
Comment (jubianchi) [inline classes/test.php]: @Grummfy no big change here, your code was already correct ;)

File: tests/units/classes/test.php [hunk 6]
```diff
@@ -1245,5 +1250,51 @@ public function testStaticMethod()
 						->isEqualTo(array($return1, $return2, $return3))
 			;
 		}
+
+		public function testNewMockInstance()
+		{
+			$this
+				->if($test = new emptyTest())
+				->then
+					->object($mock = $test->newMockInstance('stdClass'))
```
Comment (jubianchi) [inline tests/units/classes/test.php]: Here we test all the possible call to `newMockInstance` and check that the mock is correct

File: tests/units/classes/test.php [hunk 7]
```diff
@@ -1245,5 +1250,51 @@ public function testStaticMethod()
...
+						->isInstanceOf('stdClass')
+						->isNotIdenticalTo($mock)
+					->object($test->newMockInstance('stdClass', 'foobar'))
+						->isInstanceOf('foobar\stdClass')
+						->isInstanceOf('stdClass')
+					->object($test->newMockInstance('stdClass', 'foo', 'bar'))
+						->isInstanceOf('foo\bar')
+						->isInstanceOf('stdClass')
+
+				->given($arguments = array($firstArgument = uniqid(), $secondArgument = rand(0, PHP_INT_MAX)))
```
Comment (jubianchi) [inline tests/units/classes/test.php]: Here we check that when we pass constructor arguments they are correctly passed to the mock constructor

File: tests/units/classes/test.php [hunk 8]
```diff
@@ -1245,5 +1250,51 @@ public function testStaticMethod()
...
+				->then
+					->object($mock = $test->newMockInstance('mageekguy\atoum\dummy', null, null, $arguments))
+						->isInstanceOf('mock\mageekguy\atoum\dummy')
+						->isInstanceOf('mageekguy\atoum\dummy')
+					->mock($mock)
+						->call('__construct')->withArguments($firstArgument, $secondArgument)->once
+
+				->given($arguments = array(uniqid(), rand(0, PHP_INT_MAX), $controller = new mock\controller()))
```
Comment (jubianchi) [inline tests/units/classes/test.php]: We add an additional argument which is a mock controller and check that atoum correctly injected it into the mock instance (via constructor arguments)

File: tests/units/classes/test.php [hunk 9]
```diff
@@ -1245,5 +1250,51 @@ public function testStaticMethod()
...
+				->then
+					->object($mock = $test->newMockInstance('mageekguy\atoum\dummy', null, null, $arguments))
+						->isInstanceOf('mock\mageekguy\atoum\dummy')
+						->isInstanceOf('mageekguy\atoum\dummy')
+					->object($mock->getMockController())->isIdenticalTo($controller)
+
+				->given(
+					$arguments = array(uniqid(), rand(0, PHP_INT_MAX)),
+					$controller = new mock\controller()
```
Comment (jubianchi) [inline tests/units/classes/test.php]: We add an additional argument which is a mock controller and check that atoum correctly injected it into the mock instance (via automagic injection)

File: tests/units/classes/test.php [hunk 10]
```diff
@@ -1245,5 +1250,51 @@ public function testStaticMethod()
 						->isEqualTo(array($return1, $return2, $return3))
 			;
 		}
+
+		public function testNewMockInstance()
+		{
+			$this
+				->if($test = new emptyTest())
```
Comment (jubianchi) [inline tests/units/classes/test.php]: we use a dummy `emptyTest` class (declared at the top of this file) as the tested class

File: classes/test.php (hunk 4 above)
Comment (Grummfy) [inline classes/test.php]: I have seen that in mockgenerator we can defined another reflexion class. Perhaps we should inject it here too?

Comment (jubianchi) [inline classes/test.php]: yes at some places we use a reflectionClassFactory (for example in the constructor of this class) to deeply test code behavior but it was not required here: those tests are tightly coupled to implementation, I prefered to only check external behaviors

**Pull request merged**